lib/install/flow.go

```diff
@@ -358,42 +359,68 @@ func (i *Installer) send(e Event) {
 }
 
-func (i *Installer) canContinue(servers []checks.ServerInfo) bool {
-	profiles := make(map[string]int)
-	for _, node := range i.flavor.Nodes {
-		profiles[node.Profile] = node.Count
-	}
-	for _, s := range servers {
-		counter := profiles[s.Role]
-		counter -= 1
-		profiles[s.Role] = counter
-	}
-	for role, counter := range profiles {
-		if counter > 0 {
-			i.send(Event{
-				Progress: &ops.ProgressEntry{
-					Message: fmt.Sprintf(
-						"Still waiting for %v nodes of role %q", counter, role),
-				},
-			})
-			log.Infof("Still waiting for %v nodes of role %q.", counter, role)
-			return false
-		}
-	}
-	i.send(Event{
-		Progress: &ops.ProgressEntry{
-			Message: "All agents have connected!",
-		},
-	})
-	log.Info("All agents have connected!")
-	return true
+// formatProfiles outputs a table with information about node profiles
+// that need to join in order for installation to proceed.
+func (i *Installer) formatProfiles(profiles map[string]int) string {
+	var buf bytes.Buffer
+	w := new(tabwriter.Writer)
+	w.Init(&buf, 0, 8, 1, '\t', 0)
+	fmt.Fprintf(w, "Role\tNodes\tCommand\n")
+	fmt.Fprintf(w, "----\t-----\t-------\n")
+	for role, nodes := range profiles {
+		fmt.Fprintf(w, "%v\t%v\t%v\n", role, nodes,
+			fmt.Sprintf("gravity join %v --token=%v --role=%v",
+				i.AdvertiseAddr, i.Token.Token, role))
+	}
+	w.Flush()
+	return buf.String()
+}
+
+// canContinue returns true if the installation can commence based on the
+// provided agent report and false if not all agents have joined yet.
+func (i *Installer) canContinue(report *ops.AgentReport) bool {
+	// See if any new nodes have joined or left since previous agent report.
+	joined, left := report.Diff(i.agentReport)
+	for _, server := range joined {
+		i.sendMessage(color.GreenString("Successfully added %q node on %v",
+			server.Role, utils.ExtractHost(server.AdvertiseAddr)))
+	}
+	for _, server := range left {
+		i.sendMessage(color.YellowString("Node %q on %v has left",
+			server.Role, utils.ExtractHost(server.AdvertiseAddr)))
+	}
+	// Save the current agent report so we can compare against it next iteration.
+	i.agentReport = report
+	// See if the current agent report satisfies the selected flavor.
+	needed, extra := report.Check(i.flavor)
+	if len(needed) == 0 && len(extra) == 0 {
+		i.sendMessage(color.GreenString("All agents have connected!"))
+		return true
+	}
+	// If there were no changes compared to previous report, do not
+	// output anything.
+	if len(joined) == 0 && len(left) == 0 {
+		return false
+	}
+	// Dump the table with remaining nodes that need to join.
+	i.sendMessage(fmt.Sprintf("Please execute the following join commands on target nodes:\n%v",
+		i.formatProfiles(needed)))
+	// If there are any extra agents with roles we don't expect for
+	// the selected flavor, they need to leave.
+	for _, server := range extra {
+		i.sendMessage(color.RedString("Node %q on %v is not a part of the flavor, shut it down",
```

Comment:
It will only be printed during initial agents cluster formation, I can see it happening that someone type in incorrect role (even with the table) so this might be useful. Arguably, a better approach would be to just shut down the agent with incorrect role but it's not implemented yet and I have to handle this scenario as well.